### PR TITLE
feat: add binary gcd algorithm

### PR DESCRIPTION
**Pull request by tolis-0**
#### Description of Change
Introduced an implementation of the binary GCD algorithm

#### Checklist
- [X] Added description of change
- [X] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [X] Added tests and example, test must pass
- [X] Relevant documentation/comments is changed or added
- [X] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [X] Search previous suggestions before making a new one, as yours may be a duplicate.
- [X] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: math/binary_gcd.c [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+#define trailing_zeros(x) __builtin_ctzll(x)
+#endif
+
+/**
+ * @brief Get greatest common divisor of u and v without division
+ * @param [in] u nonnegative integer
+ * @param [in] v nonnegative integer
+ * @return gcd(u,v) in unsigned long type
+ */
+unsigned long binary_gcd(unsigned long u, unsigned long v)
```
Comment (realstealthninja) [inline math/binary_gcd.c:26]:
```suggestion
uint64_t binary_gcd(uint64_t u, uint64_t v)
```

File: math/binary_gcd.c [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+{
+    int shift, u_tz, v_tz;
+    unsigned long tmp;
```
Comment (realstealthninja) [inline math/binary_gcd.c:29]:
```suggestion
    uint64_t tmp;
```

File: math/binary_gcd.c [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+ * @brief Compute greatest common divisor using Binary GCD (Stein's algorithm)
+ * https://en.wikipedia.org/wiki/Binary_GCD_algorithm
+ * @author [Apostolos Pappas](https://github.com/tolis-0/)
+ * @date 28 November, 2023
+ */
+
+#include <assert.h>
+#include <locale.h>
+#include <stdio.h>
+#include <stdlib.h>
```
Comment (realstealthninja) [inline math/binary_gcd.c:12]:
```suggestion
#include <stdlib.h>
#include <stdint.h>
```

Comment (realstealthninja) [reply math/binary_gcd.c:12]: dont recall if this is necessary @Panquesito7

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by tolis-0** [comment]
@realstealthninja I committed the changes.

**Pull request closed without merging**